--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -275,7 +_,8 @@
     private @Nullable CommandStorage commandStorage;
     private final CustomBossEvents customBossEvents;
     private final RandomSequences randomSequences;
-    // Paper - per-level WeatherData
+    private final WeatherData weatherData; // Paper - per-level WeatherData // KioCG - per world weather config
+    public int weatherTick; // KioCG - per world weather config: 全局天气每游戏刻只计算一次
     private final ServerFunctionManager functionManager;
     private boolean enforceWhitelist;
     private boolean usingWhitelist;
@@ -550,7 +_,7 @@
         this.storageSource = storageSource;
         this.playerDataStorage = storageSource.createPlayerStorage();
         this.randomSequences = this.savedDataStorage.computeIfAbsent(RandomSequences.TYPE);
-        // Paper - per-level WeatherData
+        this.weatherData = this.getDataStorage().computeIfAbsent(WeatherData.TYPE); // Paper - per-level WeatherData // KioCG - per world weather config
         // Paper - per-level GameRules
         this.fixerUpper = fixerUpper;
         this.functionManager = new ServerFunctionManager(this, this.resources.managers.getFunctionLibrary());
@@ -1141,6 +_,8 @@
         this.safeShutdown(wait, false);
     }
     public void safeShutdown(final boolean wait, final boolean isRestarting) {
+        com.kiocg.server.tasks.BossBarTask.stopAll(); // KioCG - bossbar task
+
         this.isRestarting = isRestarting;
         this.hasLoggedStop = true; // Paper - Debugging
         if (this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
@@ -1638,6 +_,17 @@
             }
         }
 
+        // KioCG start - chunk load
+        if (this.tickCount % 20 == 0) {
+            for (ServerLevel serverLevel : this.getAllLevels()) {
+                for (net.minecraft.world.level.chunk.LevelChunk chunk : serverLevel.moonrise$getTickingChunks()) {
+                    chunk.getChunkLoad().newTick();
+                }
+            }
+            com.kiocg.world.level.chunk.ChunkLoad.nextTick();
+        }
+        // KioCG end - chunk load
+
         this.server.spark.tickStart(); // Paper - spark
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
         this.tickCount++;
@@ -1809,13 +_,13 @@
         if (this.tickRateManager.runsNormally()) {
             profiler.push("clocks");
             // Paper start - per-world time
-            if (io.papermc.paper.configuration.GlobalConfiguration.get().time.affectsAllWorlds) {
-                this.clockManager.tick();
-            } else {
-                for (ServerLevel level : this.getAllLevels()) {
+            // KioCG start - per world time config
+            this.clockManager.tick();
+            for (ServerLevel level : this.getAllLevels()) {
+                if (level.paperConfig().kiocg.perWorldTime)
                     level.clockManager().tick();
-                }
             }
+            // KioCG end - per world time config
             // Paper end - per-world time
             profiler.pop();
         }
@@ -2595,9 +_,9 @@
         weatherData.setThundering(thundering, org.bukkit.event.weather.ThunderChangeEvent.Cause.COMMAND); // Paper - per-level WeatherData
     }
 
-    @Deprecated(forRemoval = true) @io.papermc.paper.annotation.DoNotUse // Paper
+    // @Deprecated(forRemoval = true) @io.papermc.paper.annotation.DoNotUse // Paper // KioCG - per world weather config
     public WeatherData getWeatherData() {
-        throw new UnsupportedOperationException("Use ServerLevel.getWeatherData() instead"); // Paper
+        return this.weatherData; // Paper // KioCG - per world weather config
     }
 
     public boolean isEnforceWhitelist() {
@@ -2945,6 +_,7 @@
             byte event = (byte)((Boolean)value ? 22 : 23);
 
             for (ServerPlayer player : level.players()) { // Paper - per-world game rules
+                if (event != net.minecraft.world.entity.EntityEvent.FULL_DEBUG_INFO && player.getBukkitEntity().hasPermission("kiocg.reduced_debug_info.bypass")) event = net.minecraft.world.entity.EntityEvent.FULL_DEBUG_INFO; // KioCG - more permission
                 player.connection.send(new ClientboundEntityEventPacket(player, event));
             }
         } else if (rule == GameRules.LIMITED_CRAFTING || rule == GameRules.IMMEDIATE_RESPAWN) {
