--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -127,5 +_,6 @@
     public void updateComponentsAndStaticRegistryTags() {
         this.postponedTags.forEach(Registry.PendingTags::apply);
         this.newComponents.forEach(DataComponentInitializers.PendingComponents::apply);
+        net.minecraft.world.item.Items.modifyItems(); // KioCG - customize default components
     }
 }
