--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryFurnace.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryFurnace.java
@@ -44,7 +_,12 @@
     @Override
     public boolean isFuel(ItemStack stack) {
         net.minecraft.server.level.ServerLevel world = ((org.bukkit.craftbukkit.CraftWorld) org.bukkit.Bukkit.getWorlds().get(0)).getHandle();
-        return stack != null && !stack.getType().isEmpty() && world.fuelValues().isFuel(CraftItemStack.asNMSCopy(stack));
+        // KioCG start - change container behavior
+        net.minecraft.world.level.block.entity.FuelValues.FuelType fuelType = this.inventory instanceof net.minecraft.world.level.block.entity.SmokerBlockEntity
+                ? net.minecraft.world.level.block.entity.FuelValues.FuelType.SMOKER
+                : net.minecraft.world.level.block.entity.FuelValues.FuelType.FURNACE;
+        return stack != null && !stack.getType().isEmpty() && world.fuelValues(fuelType).isFuel(CraftItemStack.asNMSCopy(stack));
+        // KioCG end - change container behavior
     }
 
     @Override
