--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -901,7 +_,10 @@
                 .limit(5L)
                 .toList();
             if (nearbyVillagersThatWantAGolem.size() >= villagersNeededToAgree) {
-                if (SpawnUtil.trySpawnMob( // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
+                // KioCG start - mob farm restrictor
+                Optional<net.minecraft.world.entity.animal.golem.IronGolem> ironGolem;
+                if ((ironGolem = SpawnUtil.trySpawnMob( // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
+                        // KioCG end - mob farm restrictor
                         EntityType.IRON_GOLEM,
                         EntitySpawnReason.MOB_SUMMONED,
                         level,
@@ -913,9 +_,10 @@
                         false,
                         org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE, // CraftBukkit,
                         () -> {GolemSensor.golemDetected(this);} // Paper - Set Golem Last Seen to stop it from spawning another one
-                    )
+                )) // KioCG - mob farm restrictor
                     .isPresent()) { // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                     nearbyVillagers.forEach(GolemSensor::golemDetected);
+                    level.increaseMobLimit(ironGolem.get(), 4); // KioCG - mob farm restrictor
                 }
             }
         }
