--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -145,6 +_,7 @@
     public @Nullable SpawnGroupData finalizeSpawn(
         final ServerLevelAccessor level, final DifficultyInstance difficulty, final EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
     ) {
+        if (this instanceof CaveSpider) return super.finalizeSpawn(level, difficulty, spawnReason, groupData); // KioCG - mob random scale
         groupData = super.finalizeSpawn(level, difficulty, spawnReason, groupData);
         RandomSource random = level.getRandom();
         if (random.nextInt(100) == 0) {
