--- a/net/minecraft/world/level/levelgen/NoiseGeneratorSettings.java
+++ b/net/minecraft/world/level/levelgen/NoiseGeneratorSettings.java
@@ -67,7 +_,7 @@
     }
 
     public boolean oreVeinsEnabled() {
-        return this.oreVeinsEnabled && !SharedConstants.DEBUG_DISABLE_ORE_VEINS;
+        return false && this.oreVeinsEnabled && !SharedConstants.DEBUG_DISABLE_ORE_VEINS; // KioCG - disable ore veins
     }
 
     public WorldgenRandom.Algorithm getRandomSource() {
