--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1891,7 +_,7 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (this.dead && killer instanceof WitherBoss) { // Paper
-                if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING_KIOCG)) { // KioCG - better mob griefing setting
                     BlockPos pos = this.blockPosition();
                     BlockState state = Blocks.WITHER_ROSE.defaultBlockState();
                     if (this.level().getBlockState(pos).isAir() && state.canSurvive(this.level(), pos)) {
@@ -2899,6 +_,17 @@
         }
     }
 
+    // KioCG start - more api
+    public @Nullable InteractionHand getHandWithItem(final Item item) {
+        if (this.getMainHandItem().is(item)) {
+            return InteractionHand.MAIN_HAND;
+        } else if (this.getOffhandItem().is(item)) {
+            return InteractionHand.OFF_HAND;
+        }
+        return null;
+    }
+    // KioCG end - more api
+
     public void setItemInHand(final InteractionHand hand, final ItemStack itemStack) {
         if (hand == InteractionHand.MAIN_HAND) {
             this.setItemSlot(EquipmentSlot.MAINHAND, itemStack);
@@ -3588,6 +_,8 @@
                 final org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(current);
                 if (this instanceof ServerPlayer && slot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) {
                     new com.destroystokyo.paper.event.player.PlayerArmorChangeEvent((org.bukkit.entity.Player)this.getBukkitEntity(), com.destroystokyo.paper.event.player.PlayerArmorChangeEvent.SlotType.valueOf(slot.name()), oldItem, newItem).callEvent();
+
+                    com.kiocg.world.item.enchantment.Insulation.update((ServerPlayer) this, slot, current); // KioCG - more enchantments
                 }
                 // Paper end - EntityEquipmentChangedEvent, PlayerArmorChangeEvent
                 if (changedItems == null) {
