--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -123,6 +_,7 @@
     private static final boolean DEFAULT_LEFT_HANDED = false;
     private static final boolean DEFAULT_NO_AI = false;
     protected static final Identifier RANDOM_SPAWN_BONUS_ID = Identifier.withDefaultNamespace("random_spawn_bonus");
+    protected static final Identifier RANDOM_SCALE_ID = Identifier.withDefaultNamespace("random_scale"); // KioCG - mob random scale
     public static final String TAG_DROP_CHANCES = "drop_chances";
     public static final String TAG_LEFT_HANDED = "LeftHanded";
     public static final String TAG_CAN_PICK_UP_LOOT = "CanPickUpLoot";
@@ -542,7 +_,7 @@
             && this.canPickUpLoot()
             && this.isAlive()
             && !this.dead
-            && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+            /*&& serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)*/) { // KioCG - better mob griefing setting
             Vec3i pickupReach = this.getPickupReach();
 
             for (ItemEntity entity : this.level()
@@ -763,6 +_,7 @@
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+            if (!this.getType().getCategory().isFriendly()) this.level().getMinecraftWorld().decreaseMobLimit(this); // KioCG - mob farm restrictor
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
             if (player != null) {
@@ -777,11 +_,13 @@
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    if (!this.getType().getCategory().isFriendly()) this.level().getMinecraftWorld().decreaseMobLimit(this); // KioCG - mob farm restrictor
                 }
 
                 if (despawnRangePair.soft().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy)) {
                     if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && this.removeWhenFarAway(distanceSquared)) {
                         this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                        if (!this.getType().getCategory().isFriendly()) this.level().getMinecraftWorld().decreaseMobLimit(this); // KioCG - mob farm restrictor
                     }
                 } else {
                     // Paper end - Configurable despawn distances
@@ -1255,6 +_,14 @@
                 new AttributeModifier(RANDOM_SPAWN_BONUS_ID, random.triangle(0.0, 0.11485000000000001), AttributeModifier.Operation.ADD_MULTIPLIED_BASE)
             );
         }
+        // KioCG start - mob random scale
+        AttributeInstance scale = Objects.requireNonNull(this.getAttribute(Attributes.SCALE));
+        if (!scale.hasModifier(RANDOM_SCALE_ID)) {
+            scale.addPermanentModifier(
+                    new AttributeModifier(RANDOM_SCALE_ID, random.triangleLeftHalf(0.0, 0.11485000000000001), AttributeModifier.Operation.ADD_MULTIPLIED_BASE)
+            );
+        }
+        // KioCG end - mob random scale
 
         this.setLeftHanded(random.nextFloat() < 0.05F);
         return groupData;
@@ -1646,6 +_,13 @@
             }
         }
     }
+
+    // KioCG start - chunk load
+    @Override
+    public boolean shouldAnalyzeLoad() {
+        return super.shouldAnalyzeLoad() && (this.isPersistenceRequired() || this.requiresCustomPersistence() || !this.removeWhenFarAway(0.0));
+    }
+    // KioCG end - chunk load
 
     @Override
     public @Nullable ItemStack getPickResult() {
