--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -270,7 +_,14 @@
     private static boolean hopperPull(final Level level, final Hopper hopper, final Container container, ItemStack origItemStack, final int i) {
         ItemStack movedItem = origItemStack;
         final int originalItemCount = origItemStack.getCount();
-        final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+        // KioCG start - change container behavior
+        final int movedItemCount;
+        if (container instanceof net.minecraft.world.level.block.ComposterBlock.OutputContainer) {
+            movedItemCount = originalItemCount;
+        } else {
+            movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
+        }
+        // KioCG end - change container behavior
         container.setChanged(); // original logic always marks source inv as changed even if no move happens.
         movedItem.setCount(movedItemCount);
 
