--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -652,7 +_,16 @@
 
     private static @Nullable Container getSourceContainer(final Level level, final Hopper hopper, final BlockPos pos, final BlockState state) {
         // Paper start
-        final Container inventory = HopperBlockEntity.getContainerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
+        // KioCG start - change container behavior
+        final Container inventory;
+        if (state.getBlock() instanceof net.minecraft.world.level.block.ComposterBlock) {
+            net.minecraft.world.level.block.ComposterBlock.hopperSuckInTrigger(state, level, pos);
+            inventory = null;
+        } else {
+            inventory = HopperBlockEntity.getContainerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
+        }
+        // KioCG end - change container behavior
+
         if (org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0) return inventory;
 
         final BlockPos hopperPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY(), hopper.getLevelZ());
