--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1153,10 +_,21 @@
             return;
         }
         List<DefaultDrop> loot = new java.util.ArrayList<>(this.getInventory().getContainerSize()); // Paper - Restore vanilla drops behavior
+        List<org.bukkit.inventory.ItemStack> itemsToKeep = new java.util.ArrayList<>(); // KioCG - death drop protection
         boolean keepInventory = this.level().getGameRules().get(GameRules.KEEP_INVENTORY) || this.isSpectator();
         if (!keepInventory) {
-            for (ItemStack item : this.getInventory().getContents()) {
+            // KioCG start - death drop protection
+            List<ItemStack> contents = this.getInventory().getContents();
+            for (int i = 0; i < contents.size(); i++) {
+                ItemStack item = contents.get(i);
+                // KioCG end - death drop protection
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, net.minecraft.world.item.enchantment.EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
+                    // KioCG start - death drop protection
+                    if (i < Inventory.SELECTION_SIZE || i >= Inventory.INVENTORY_SIZE) {
+                        itemsToKeep.add(item.asBukkitMirror());
+                        continue;
+                    }
+                    // KioCG end - death drop protection
                     loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false, null))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
                 }
             }
@@ -1171,8 +_,8 @@
 
         Component defaultMessage = this.getCombatTracker().getDeathMessage();
 
-        this.keepLevel = keepInventory; // SPIGOT-2222: pre-set keepLevel
-        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, source, loot, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), showDeathMessage, keepInventory); // Paper - Adventure; Expand PlayerDeathEvent API
+        this.keepLevel = true; // SPIGOT-2222: pre-set keepLevel // KioCG - EXP is money: 玩家死亡不要掉落经验
+        org.bukkit.event.entity.PlayerDeathEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerDeathEvent(this, source, loot, itemsToKeep, io.papermc.paper.adventure.PaperAdventure.asAdventure(defaultMessage), showDeathMessage, keepInventory); // Paper - Adventure; Expand PlayerDeathEvent API // KioCG - death drop protection
         // Paper start - cancellable death event
         if (event.isCancelled()) {
             // make compatible with plugins that might have already set the health in an event listener
