--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -221,6 +_,8 @@
     private final boolean tickTime;
     private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
 
+    private final Object2IntOpenHashMap<ChunkPos> mobFarmRestrictor = new Object2IntOpenHashMap<>(); // KioCG - mob farm restrictor
+
     // CraftBukkit start
     private final ResourceKey<LevelStem> typeKey;
     public final String bukkitName;
@@ -683,7 +_,7 @@
         this.chunkSource.getGeneratorState().ensureStructuresGenerated();
         this.portalForcer = new PortalForcer(this);
         if (this.canHaveWeather()) {
-            this.prepareWeather(this.weatherData); // Paper - per-level WeatherData
+            this.prepareWeather(this.paperConfig().kiocg.perWorldWeather ? this.weatherData : server.getWeatherData()); // Paper - per-level WeatherData // KioCG - per world weather config
         }
 
         this.raids = this.getDataStorage().computeIfAbsent(Raids.TYPE);
@@ -817,6 +_,13 @@
             this.tickTime();
         }
 
+        // KioCG start - mob farm restrictor
+        long time = this.getOverworldClockTime() % net.minecraft.SharedConstants.TICKS_PER_GAME_DAY;
+        if (time == 13000) {
+            this.resetMobLimit();
+        }
+        // KioCG end - mob farm restrictor
+
         profiler.push("tickPending");
         if (!this.isDebug() && runs) {
             long tick = this.getGameTime();
@@ -1193,6 +_,10 @@
         if (this.canHaveWeather()) {
             WeatherData weatherData = this.getWeatherData();
             if (this.getGameRules().get(GameRules.ADVANCE_WEATHER)) {
+                // KioCG start - per world weather config: 全局天气每游戏刻只计算一次
+                if (weatherData != server.getWeatherData() || server.weatherTick != server.getTickCount()) {
+                    if (weatherData == server.getWeatherData()) server.weatherTick = server.getTickCount();
+                    // KioCG end - per world weather config: 全局天气每游戏刻只计算一次
                 int clearWeatherTime = weatherData.getClearWeatherTime();
                 int thunderTime = weatherData.getThunderTime();
                 int rainTime = weatherData.getRainTime();
@@ -1231,6 +_,7 @@
                 weatherData.setClearWeatherTime(clearWeatherTime);
                 weatherData.setThundering(thundering, org.bukkit.event.weather.ThunderChangeEvent.Cause.NATURAL); // Paper - Add cause to Weather/ThunderChangeEvents
                 weatherData.setRaining(raining, org.bukkit.event.weather.WeatherChangeEvent.Cause.NATURAL); // Paper - Add cause to Weather/ThunderChangeEvents
+                } // KioCG - per world weather config: 全局天气每游戏刻只计算一次
             }
 
             this.oThunderLevel = this.thunderLevel;
@@ -2523,6 +_,41 @@
         return this.getEntities().getAll();
     }
 
+    // KioCG start - mob farm restrictor
+    public void increaseMobLimit(final Entity entity) {
+        this.mobFarmRestrictor.addTo(entity.chunkPosition(), 1);
+    }
+
+    public void increaseMobLimit(final Entity entity, final int incr) {
+        this.mobFarmRestrictor.addTo(entity.chunkPosition(), incr);
+    }
+
+    public void decreaseMobLimit(final Entity entity) {
+        if (entity.origin == null || entity.originWorld == null) {
+            return;
+        }
+
+        ChunkPos chunkPos = ChunkPos.containing(BlockPos.containing(entity.origin));
+        if (entity.originWorld.equals(this.uuid)) {
+            this.mobFarmRestrictor.addTo(chunkPos, -1);
+        } else {
+            for (ServerLevel serverLevel : this.getServer().getAllLevels()) {
+                if (serverLevel.uuid.equals(entity.originWorld)) {
+                    serverLevel.mobFarmRestrictor.addTo(chunkPos, 1);
+                }
+            }
+        }
+    }
+
+    public boolean reachMobLimit(final ChunkPos chunkPos) {
+        return this.mobFarmRestrictor.getInt(chunkPos) >= 16;
+    }
+
+    public void resetMobLimit() {
+        this.mobFarmRestrictor.clear();
+    }
+    // KioCG end - mob farm restrictor
+
     @Override
     public String toString() {
         return "ServerLevel[" + this.serverLevelData.getLevelName() + "]";
@@ -2542,7 +_,7 @@
     }
 
     public WeatherData getWeatherData() {
-        return this.weatherData; // Paper - per-level WeatherData
+        return this.paperConfig().kiocg.perWorldWeather ? this.weatherData : this.server.getWeatherData(); // Paper - per-level WeatherData // KioCG - per world weather config
     }
 
     @Override
