--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -221,6 +_,8 @@
     private final boolean tickTime;
     private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
 
+    private final Object2IntOpenHashMap<ChunkPos> mobFarmRestrictor = new Object2IntOpenHashMap<>(); // KioCG - mob farm restrictor
+
     // CraftBukkit start
     private final ResourceKey<LevelStem> typeKey;
     public final String bukkitName;
@@ -690,7 +_,7 @@
         this.chunkSource.getGeneratorState().ensureStructuresGenerated();
         this.portalForcer = new PortalForcer(this);
         if (this.canHaveWeather()) {
-            this.prepareWeather(this.weatherData); // Paper - per-level WeatherData
+            this.prepareWeather(this.paperConfig().kiocg.perWorldWeather ? this.weatherData : server.getWeatherData()); // Paper - per-level WeatherData // KioCG - per world weather config
         }
 
         this.raids = this.getDataStorage().computeIfAbsent(Raids.TYPE);
@@ -822,6 +_,27 @@
         this.updateSkyBrightness();
         if (runs) {
             this.tickTime();
+
+            // KioCG start - mob farm restrictor
+            long time = this.getOverworldClockTime() % net.minecraft.SharedConstants.TICKS_PER_GAME_DAY;
+            if (time == 13000) {
+                this.resetMobLimit();
+            }
+            // KioCG end - mob farm restrictor
+
+            // KioCG start - keep block destruction
+            // 客户端会清理400刻之前的方块破坏纹理
+            if (this.lagCompensationTick % 20 == 0) {
+                java.util.Iterator<BlockDestructionProgress> iterator = this.destroyingBlocks.values().iterator();
+                while (iterator.hasNext()) {
+                    BlockDestructionProgress blockDestructionProgress = iterator.next();
+                    int i = blockDestructionProgress.getUpdatedRenderTick();
+                    if (this.lagCompensationTick - i > 400) {
+                        iterator.remove();
+                    }
+                }
+            }
+            // KioCG end - keep block destruction
         }
 
         profiler.push("tickPending");
@@ -1203,6 +_,10 @@
         if (this.canHaveWeather()) {
             WeatherData weatherData = this.getWeatherData();
             if (this.getGameRules().get(GameRules.ADVANCE_WEATHER)) {
+                // KioCG start - per world weather config: 全局天气每游戏刻只计算一次
+                if (weatherData != server.getWeatherData() || server.weatherTick != server.getTickCount()) {
+                    if (weatherData == server.getWeatherData()) server.weatherTick = server.getTickCount();
+                    // KioCG end - per world weather config: 全局天气每游戏刻只计算一次
                 int clearWeatherTime = weatherData.getClearWeatherTime();
                 int thunderTime = weatherData.getThunderTime();
                 int rainTime = weatherData.getRainTime();
@@ -1241,6 +_,7 @@
                 weatherData.setClearWeatherTime(clearWeatherTime);
                 weatherData.setThundering(thundering, org.bukkit.event.weather.ThunderChangeEvent.Cause.NATURAL); // Paper - Add cause to Weather/ThunderChangeEvents
                 weatherData.setRaining(raining, org.bukkit.event.weather.WeatherChangeEvent.Cause.NATURAL); // Paper - Add cause to Weather/ThunderChangeEvents
+                } // KioCG - per world weather config: 全局天气每游戏刻只计算一次
             }
 
             this.oThunderLevel = this.thunderLevel;
@@ -1386,10 +_,20 @@
         profiler.push(entity.typeHolder()::getRegisteredName);
         profiler.incrementCounter("tickNonPassenger");
         final boolean isActive = io.papermc.paper.entity.activation.ActivationRange.checkIfActive(entity); // Paper - EAR 2
+            // KioCG start - chunk load
+            LevelChunk levelChunk = entity.shouldAnalyzeLoad() ? this.getChunkIfLoaded(entity.blockPosition()) : null;
+            if (levelChunk != null) levelChunk.getChunkLoad().startTicking();
+            try {
+            // KioCG end - chunk load
         if (isActive) { // Paper - EAR 2
         entity.tick();
         entity.postTick(); // CraftBukkit
         } else {entity.inactiveTick();} // Paper - EAR 2
+            // KioCG start - chunk load
+            } finally {
+                if (levelChunk != null) levelChunk.getChunkLoad().endTickingAndCount();
+            }
+            // KioCG end - chunk load
         profiler.pop();
 
         for (Entity passenger : entity.getPassengers()) {
@@ -1414,6 +_,11 @@
             ProfilerFiller profiler = Profiler.get();
             profiler.push(entity.typeHolder()::getRegisteredName);
             profiler.incrementCounter("tickPassenger");
+            // KioCG start - chunk load
+            LevelChunk levelChunk = entity.shouldAnalyzeLoad() ? this.getChunkIfLoaded(entity.blockPosition()) : null;
+            if (levelChunk != null) levelChunk.getChunkLoad().startTicking();
+            try {
+            // KioCG end - chunk load
             // Paper start - EAR 2
             if (isActive) {
             entity.rideTick();
@@ -1425,6 +_,11 @@
                 vehicle.positionRider(entity);
             }
             // Paper end - EAR 2
+            // KioCG start - chunk load
+            } finally {
+                if (levelChunk != null) levelChunk.getChunkLoad().endTickingAndCount();
+            }
+            // KioCG end - chunk load
             profiler.pop();
 
             for (Entity passenger : entity.getPassengers()) {
@@ -1702,6 +_,17 @@
 
     @Override
     public void destroyBlockProgress(final int id, final BlockPos blockPos, final int progress) {
+        // KioCG start - keep block destruction
+        BlockDestructionProgress blockDestructionProgress;
+        if (progress > -1) {
+            blockDestructionProgress = this.destroyingBlocks.computeIfAbsent(blockPos, _ -> new BlockDestructionProgress(BlockDestructionProgress.getNextBlockDestroyId(), blockPos));
+            blockDestructionProgress.setProgress(progress);
+            blockDestructionProgress.updateTick((int) this.lagCompensationTick);
+        } else {
+            blockDestructionProgress = this.destroyingBlocks.remove(blockPos);
+        }
+        // KioCG end - keep block destruction
+
         // CraftBukkit start
         Player breakerPlayer = null;
         Entity entity = this.getEntity(id);
@@ -1729,7 +_,7 @@
                         continue;
                     }
                     // CraftBukkit end
-                    player.connection.send(new ClientboundBlockDestructionPacket(id, blockPos, progress));
+                    player.connection.send(new ClientboundBlockDestructionPacket(breakerPlayer != null ? id : blockDestructionProgress.getId(), blockPos, progress)); // KioCG - keep block destruction
                 }
             }
         }
@@ -1989,7 +_,7 @@
         Explosion.BlockInteraction blockInteraction = switch (interactionType) {
             case NONE -> Explosion.BlockInteraction.KEEP;
             case BLOCK -> this.getDestroyType(GameRules.BLOCK_EXPLOSION_DROP_DECAY);
-            case MOB -> this.getGameRules().get(GameRules.MOB_GRIEFING)
+            case MOB -> this.getGameRules().get(GameRules.MOB_GRIEFING_KIOCG) // KioCG - better mob griefing setting
                 ? this.getDestroyType(GameRules.MOB_EXPLOSION_DROP_DECAY)
                 : Explosion.BlockInteraction.KEEP;
             case TNT -> this.getDestroyType(GameRules.TNT_EXPLOSION_DROP_DECAY);
@@ -2532,6 +_,41 @@
         return this.getEntities().getAll();
     }
 
+    // KioCG start - mob farm restrictor
+    public void increaseMobLimit(final Entity entity) {
+        this.mobFarmRestrictor.addTo(entity.chunkPosition(), 1);
+    }
+
+    public void increaseMobLimit(final Entity entity, final int incr) {
+        this.mobFarmRestrictor.addTo(entity.chunkPosition(), incr);
+    }
+
+    public void decreaseMobLimit(final Entity entity) {
+        if (entity.origin == null || entity.originWorld == null) {
+            return;
+        }
+
+        ChunkPos chunkPos = ChunkPos.containing(BlockPos.containing(entity.origin));
+        if (entity.originWorld.equals(this.uuid)) {
+            this.mobFarmRestrictor.addTo(chunkPos, -1);
+        } else {
+            for (ServerLevel serverLevel : this.getServer().getAllLevels()) {
+                if (serverLevel.uuid.equals(entity.originWorld)) {
+                    serverLevel.mobFarmRestrictor.addTo(chunkPos, 1);
+                }
+            }
+        }
+    }
+
+    public boolean reachMobLimit(final ChunkPos chunkPos) {
+        return this.mobFarmRestrictor.getInt(chunkPos) >= 16;
+    }
+
+    public void resetMobLimit() {
+        this.mobFarmRestrictor.clear();
+    }
+    // KioCG end - mob farm restrictor
+
     @Override
     public String toString() {
         return "ServerLevel[" + this.serverLevelData.getLevelName() + "]";
@@ -2551,7 +_,7 @@
     }
 
     public WeatherData getWeatherData() {
-        return this.weatherData; // Paper - per-level WeatherData
+        return this.paperConfig().kiocg.perWorldWeather ? this.weatherData : this.server.getWeatherData(); // Paper - per-level WeatherData // KioCG - per world weather config
     }
 
     @Override
