--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -48,6 +_,7 @@
     private int lastSentState = -1;
     public boolean captureSentBlockEntities = false; // Paper - Send block entities after destroy prediction
     public boolean capturedBlockEntity = false; // Paper - Send block entities after destroy prediction
+    private final List<com.kiocg.world.item.enchantment.Veinchain> veinchains = new java.util.ArrayList<>(); // KioCG - more enchantments
 
     public ServerPlayerGameMode(final ServerPlayer player) {
         this.player = player;
@@ -154,6 +_,8 @@
                 this.incrementDestroyProgress(blockState, this.destroyPos, this.destroyProgressStart);
             }
         }
+
+        this.veinchains.removeIf(veinchain -> !veinchain.tick(this.player, this.player.getMainHandItem())); // KioCG - more enchantments
     }
 
     private float incrementDestroyProgress(final BlockState blockState, final BlockPos delayedDestroyPos, final int destroyStartTick) {
@@ -227,7 +_,7 @@
                     return;
                 }
 
-                this.destroyProgressStart = this.gameTicks;
+                // this.destroyProgressStart = this.gameTicks; // KioCG - fix block destroy: moved down: 在破坏方块过程中, 又破坏了一个立即破裂的方块, 则不要重置此过程时间
                 float progress = 1.0F;
                 BlockState blockState = this.level.getBlockState(pos);
                 if (event.useInteractedBlock() != org.bukkit.event.Event.Result.DENY && !blockState.isAir()) { // Paper
@@ -265,6 +_,8 @@
                 if (!blockState.isAir() && progress >= 1.0F) {
                     this.destroyAndAck(pos, sequence, "insta mine");
                 } else {
+                    this.destroyProgressStart = this.gameTicks; // KioCG - fix block destroy: moved down: 在破坏方块过程中, 又破坏了一个立即破裂的方块, 则不要重置此过程时间
+
                     if (this.isDestroyingBlock) {
                         this.player.connection.send(new ClientboundBlockUpdatePacket(this.destroyPos, this.level.getBlockState(this.destroyPos)));
                         this.debugLogging(pos, false, sequence, "abort destroying since another started (client insta mine, server disagreed)");
@@ -333,6 +_,12 @@
     }
 
     public boolean destroyBlock(final BlockPos pos) {
+        // KioCG start - more enchantments
+        return destroyBlock(pos, true);
+    }
+
+    public boolean destroyBlock(final BlockPos pos, boolean applyChainEnchantments) {
+        // KioCG end - more enchantments
         BlockState state = this.level.getBlockState(pos);
         // CraftBukkit start - fire BlockBreakEvent
         org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, pos);
@@ -435,6 +_,16 @@
         }
         // Paper end - Trigger bee_nest_destroyed trigger in the correct place
         // CraftBukkit end
+
+        // KioCG start - more enchantments
+        applyChainEnchantments &= !player.isShiftKeyDown();
+        if (applyChainEnchantments && destroyedWith != null && changed && canDestroy) {
+            com.kiocg.world.item.enchantment.Veinchain veinchain = com.kiocg.world.item.enchantment.Veinchain.onDestroyBlock(this.level, pos, state, this.player.getMainHandItem());
+            if (veinchain != null) {
+                this.veinchains.add(veinchain);
+            }
+        }
+        // KioCG end - more enchantments
 
         return true;
     }
