--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -315,7 +_,7 @@
     public boolean hurtServer(final ServerLevel level, final DamageSource source, final float damage) {
         if (this.isRemoved()) {
             return false;
-        } else if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
+        } else if (!level.getGameRules().get(GameRules.MOB_GRIEFING_KIOCG) && source.getEntity() instanceof Player || source.getEntity() instanceof Mob) { // KioCG - better mob griefing setting
             return false;
         } else if (source.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
             // CraftBukkit start
