--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -328,7 +_,7 @@
             return false;
         }
 
-        if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
+        if (!level.getGameRules().get(GameRules.MOB_GRIEFING_KIOCG) && source.getEntity() instanceof Player || source.getEntity() instanceof Mob) { // KioCG - better mob griefing setting
             return false;
         }
 
