--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -63,7 +_,7 @@
 
                     return InteractionResult.FAIL;
                 } else {
-                    if (level.paperConfig().entities.spawning.disableMobSpawnerSpawnEggTransformation) return InteractionResult.FAIL; // Paper - Allow disabling mob spawner spawn egg transformation
+                    if (context.getPlayer() == null || !context.getPlayer().canUseGameMasterBlocks()) return InteractionResult.FAIL; // Paper - Allow disabling mob spawner spawn egg transformation // KioCG - change configuration options
                     spawnerHolder.setEntityId(type, level.getRandom());
                     level.sendBlockUpdated(pos, blockState, blockState, Block.UPDATE_ALL);
                     level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, pos);
