--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -23,6 +_,7 @@
     public static final String REALMS_NAMESPACE = "realms";
     public static final String ALLOWED_NAMESPACE_CHARACTERS = "[a-z0-9_.-]";
     public static final String PAPER_NAMESPACE = "paper"; // Paper
+    public static final String KIOCG_NAMESPACE = "kiocg"; // KioCG
     private final String namespace;
     private final String path;
 
