--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -143,6 +_,7 @@
     private final DamageSources damageSources;
     private final PalettedContainerFactory palettedContainerFactory;
     private long subTickCount;
+    public final it.unimi.dsi.fastutil.objects.Object2ObjectMap<BlockPos, net.minecraft.server.level.BlockDestructionProgress> destroyingBlocks = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(); // KioCG - keep block destruction
 
     // CraftBukkit start
     public final io.papermc.paper.antixray.ChunkPacketBlockController chunkPacketBlockController; // Paper - Anti-Xray
@@ -873,6 +_,7 @@
         this.damageSources = new DamageSources(registryAccess);
         this.entityLookup = new ca.spottedleaf.moonrise.patches.chunk_system.level.entity.dfl.DefaultEntityLookup(this); // Paper - rewrite chunk system
         this.chunkPacketBlockController = this.paperConfig().anticheat.antiXray.enabled ? new io.papermc.paper.antixray.ChunkPacketBlockControllerAntiXray(this, executor) : io.papermc.paper.antixray.ChunkPacketBlockController.NO_OPERATION_INSTANCE; // Paper - Anti-Xray
+        this.destroyingBlocks.defaultReturnValue(new net.minecraft.server.level.BlockDestructionProgress(-1, BlockPos.ZERO)); // KioCG - keep block destruction
     }
 
     // Paper start - Cancel hit for vanished players
@@ -1049,6 +_,8 @@
 
     @Override
     public boolean setBlock(final BlockPos pos, final BlockState blockState, @Block.UpdateFlags final int updateFlags, final int updateLimit) {
+        if (this.destroyingBlocks.containsKey(pos)) this.destroyBlockProgress(-1, pos, -1); // KioCG - keep block destruction
+
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
             // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
@@ -2146,7 +_,7 @@
 
     public abstract PotionBrewing potionBrewing();
 
-    public abstract FuelValues fuelValues();
+    public abstract FuelValues fuelValues(final FuelValues.FuelType fuelType); // KioCG - change container behavior
 
     public int getClientLeafTintColor(final BlockPos pos) {
         return 0;
