--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -184,7 +_,7 @@
         @Comment("This setting controls if the vanilla damage tick should be skipped if damage was blocked via a shield.")
         public boolean skipVanillaDamageTickWhenShieldBlocked = false;
         @Comment("This setting controls what compression format is used for region files.")
-        public CompressionFormat compressionFormat = CompressionFormat.ZLIB;
+        public CompressionFormat compressionFormat = CompressionFormat.LZ4; // KioCG - change configuration options
         @Comment("This setting controls if equipment should be updated when handling certain player actions.")
         public boolean updateEquipmentOnPlayerActions = true;
         @Comment("This setting controls what item data components don't need to be sanitized in oversized item obfuscation. Adding them re-enables exploits, but may be needed for certain resource packs. (Expected: minecraft:container, minecraft:charged_projectiles and minecraft:bundle_contents)")
@@ -403,4 +_,19 @@
     public class UpdateChecker extends ConfigurationPart {
         public boolean enabled = true;
     }
+
+    // KioCG start - kiocg config
+    public Kiocg kiocg;
+
+    public class Kiocg extends ConfigurationPart {
+        public Player player;
+
+        public class Player extends ConfigurationPart {
+            public int initialLevel = 60 * 30;
+            public Component permadeath = Component.text("因电荷消耗殆尽被系统抹除数据", NamedTextColor.RED);
+            public Component permadeathWarningTitle = Component.text("电荷已消耗殆尽", NamedTextColor.GRAY);
+            public Component permadeathWarningSubtitle = Component.text("你的下一次死亡将是永久的", NamedTextColor.RED);
+        }
+    }
+    // KioCG end - kiocg config
 }
