--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -587,4 +_,13 @@
         	HORIZONTAL_FIRST_OUTWARD, HORIZONTAL_FIRST_INWARD, VERTICAL_FIRST_OUTWARD, VERTICAL_FIRST_INWARD
         }
     }
+
+    // KioCG start - kiocg config
+    public Kiocg kiocg;
+
+    public class Kiocg extends ConfigurationPart {
+        public boolean perWorldTime = false; // TODO - Paper尚未实现独立世界时间
+        public boolean perWorldWeather = false;
+    }
+    // KioCG end - kiocg config
 }
