--- a/net/minecraft/world/item/crafting/display/SlotDisplayContext.java
+++ b/net/minecraft/world/item/crafting/display/SlotDisplayContext.java
@@ -13,6 +_,6 @@
     public static final ContextKeySet CONTEXT = new ContextKeySet.Builder().optional(FUEL_VALUES).optional(REGISTRIES).build();
 
     public static ContextMap fromLevel(final Level level) {
-        return new ContextMap.Builder().withParameter(FUEL_VALUES, level.fuelValues()).withParameter(REGISTRIES, level.registryAccess()).create(CONTEXT);
+        return new ContextMap.Builder().withParameter(FUEL_VALUES, level.fuelValues(null)).withParameter(REGISTRIES, level.registryAccess()).create(CONTEXT); // KioCG - change container behavior
     }
 }
