--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -240,7 +_,7 @@
     @Override
     protected void onPlace(final BlockState state, final Level level, final BlockPos pos, final BlockState oldState, final boolean movedByPiston) {
         if (state.getValue(LEVEL) == 7) {
-            level.scheduleTick(pos, state.getBlock(), 20);
+            level.scheduleTick(pos, state.getBlock(), 20 * 5 * 60); // KioCG - change container behavior
         }
     }
 
@@ -323,7 +_,7 @@
         // CraftBukkit end
         if (!level.isClientSide()) {
             Vec3 itemPos = Vec3.atLowerCornerWithOffset(pos, 0.5, 1.01, 0.5).offsetRandomXZ(level.getRandom(), 0.7F);
-            ItemEntity entity = new ItemEntity(level, itemPos.x(), itemPos.y(), itemPos.z(), new ItemStack(Items.BONE_MEAL));
+            ItemEntity entity = new ItemEntity(level, itemPos.x(), itemPos.y(), itemPos.z(), new ItemStack(Items.BONE_MEAL, level.getRandom().nextInt(16, 25))); // KioCG - change container behavior
             entity.setDefaultPickUpDelay();
             level.addFreshEntity(entity);
         }
@@ -382,7 +_,7 @@
         level.setBlock(pos, newState, Block.UPDATE_ALL);
         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(sourceEntity, newState));
         if (newLevel == 7) {
-            level.scheduleTick(pos, state.getBlock(), 20);
+            level.scheduleTick(pos, state.getBlock(), 20 * 5 * 60); // KioCG - change container behavior
         }
 
         return newState;
@@ -420,7 +_,7 @@
     public WorldlyContainer getContainer(final BlockState state, final LevelAccessor level, final BlockPos pos) {
         int contentLevel = state.getValue(LEVEL);
         if (contentLevel == 8) {
-            return new ComposterBlock.OutputContainer(state, level, pos, new ItemStack(Items.BONE_MEAL));
+            return new ComposterBlock.OutputContainer(state, level, pos, new ItemStack(Items.BONE_MEAL, level.getRandom().nextInt(16, 25))); // KioCG - change container behavior
         } else {
             return contentLevel < 7 ? new ComposterBlock.InputContainer(state, level, pos) : new ComposterBlock.EmptyContainer(level, pos); // CraftBukkit - empty levelAccessor, blockPos
         }
@@ -504,6 +_,7 @@
         private final LevelAccessor level;
         private final BlockPos pos;
         private boolean changed;
+        private final int originalCount; // KioCG - change container behavior
 
         public OutputContainer(final BlockState state, final LevelAccessor level, final BlockPos pos, final ItemStack contents) {
             super(contents);
@@ -511,6 +_,7 @@
             this.state = state;
             this.level = level;
             this.pos = pos;
+            this.originalCount = contents.getCount(); // KioCG - change container behavior
         }
 
         @Override
@@ -536,7 +_,7 @@
         @Override
         public void setChanged() {
             // CraftBukkit start - allow putting items back (eg cancelled InventoryMoveItemEvent)
-            if (this.isEmpty()) {
+            if (this.items.getFirst().getCount() < this.originalCount) { // KioCG - change container behavior
             ComposterBlock.empty(null, this.state, this.level, this.pos);
             this.changed = true;
             } else {
