--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -240,7 +_,7 @@
     @Override
     protected void onPlace(final BlockState state, final Level level, final BlockPos pos, final BlockState oldState, final boolean movedByPiston) {
         if (state.getValue(LEVEL) == 7) {
-            level.scheduleTick(pos, state.getBlock(), 20);
+            level.scheduleTick(pos, state.getBlock(), 20 * 5 * 60); // KioCG - change container behavior
         }
     }
 
@@ -323,7 +_,7 @@
         // CraftBukkit end
         if (!level.isClientSide()) {
             Vec3 itemPos = Vec3.atLowerCornerWithOffset(pos, 0.5, 1.01, 0.5).offsetRandomXZ(level.getRandom(), 0.7F);
-            ItemEntity entity = new ItemEntity(level, itemPos.x(), itemPos.y(), itemPos.z(), new ItemStack(Items.BONE_MEAL));
+            ItemEntity entity = new ItemEntity(level, itemPos.x(), itemPos.y(), itemPos.z(), new ItemStack(Items.BONE_MEAL, level.getRandom().nextInt(16, 25))); // KioCG - change container behavior
             entity.setDefaultPickUpDelay();
             level.addFreshEntity(entity);
         }
@@ -333,6 +_,21 @@
         return emptyState;
     }
 
+    // KioCG start - change container behavior
+    public static void hopperSuckInTrigger(final BlockState state, final Level level, final BlockPos pos) {
+        if (state.getValue(LEVEL) == 8) {
+            if (!level.isClientSide()) {
+                Vec3 itemPos = Vec3.atCenterOf(pos);
+                ItemEntity entity = new ItemEntity(level, itemPos.x(), itemPos.y(), itemPos.z(), new ItemStack(Items.BONE_MEAL, level.getRandom().nextInt(16, 25)));
+                entity.setDefaultPickUpDelay();
+                level.addFreshEntity(entity);
+            }
+
+            empty(null, state, level, pos);
+        }
+    }
+    // KioCG end - change container behavior
+
     private static BlockState empty(final @Nullable Entity sourceEntity, final BlockState state, final LevelAccessor level, final BlockPos pos) {
         BlockState newState = state.setValue(LEVEL, 0);
         level.setBlock(pos, newState, Block.UPDATE_ALL);
@@ -381,7 +_,7 @@
             level.setBlock(pos, newState, Block.UPDATE_ALL);
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(sourceEntity, newState));
             if (newLevel == 7) {
-                level.scheduleTick(pos, state.getBlock(), 20);
+                level.scheduleTick(pos, state.getBlock(), 20 * 5 * 60); // KioCG - change container behavior
             }
 
             return newState;
@@ -419,7 +_,7 @@
     @Override
     public WorldlyContainer getContainer(final BlockState state, final LevelAccessor level, final BlockPos pos) {
         int contentLevel = state.getValue(LEVEL);
-        if (contentLevel == 8) {
+        if (false && contentLevel == 8) { // KioCG - change container behavior
             return new ComposterBlock.OutputContainer(state, level, pos, new ItemStack(Items.BONE_MEAL));
         } else {
             return (WorldlyContainer)(contentLevel < 7 ? new ComposterBlock.InputContainer(state, level, pos) : new ComposterBlock.EmptyContainer(level, pos)); // CraftBukkit - empty levelAccessor, blockPos
