--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -124,7 +_,7 @@
                 BlockState blockState = level.getBlockState(below.offset(xx, 0, zz));
                 if (blockState.is(BlockTags.GROWS_CROPS)) {
                     blockSpeed = 1.0F;
-                    if (blockState.getValueOrElse(FarmlandBlock.MOISTURE, 0) > 0) {
+                    if (blockState.getValueOrElse(FarmlandBlock.MOISTURE, 0) == FarmlandBlock.MAX_MOISTURE) { // KioCG - irrigate farmland
                         blockSpeed = 3.0F;
                     }
                 }
@@ -177,7 +_,7 @@
         final boolean isPrecise
     ) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !serverLevel.getGameRules().get(GameRules.MOB_GRIEFING))) { // CraftBukkit
+        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !serverLevel.getGameRules().get(GameRules.MOB_GRIEFING_KIOCG))) { // CraftBukkit // KioCG - better mob griefing setting
             serverLevel.destroyBlock(pos, true, entity);
         }
 
