--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -174,7 +_,7 @@
         playerConnection.suspendFlushing();
         GameRules gameRules = level.getGameRules();
         boolean immediateRespawn = gameRules.get(GameRules.IMMEDIATE_RESPAWN);
-        boolean reducedDebugInfo = gameRules.get(GameRules.REDUCED_DEBUG_INFO);
+        boolean reducedDebugInfo = gameRules.get(GameRules.REDUCED_DEBUG_INFO) && !player.getBukkitEntity().hasPermission("kiocg.reduced_debug_info.bypass"); // KioCG - more permission
         boolean doLimitedCrafting = gameRules.get(GameRules.LIMITED_CRAFTING);
         playerConnection.send(
             new ClientboundLoginPacket(
@@ -432,6 +_,9 @@
     }
     public net.kyori.adventure.text.@Nullable Component remove(final ServerPlayer player, final net.kyori.adventure.text.Component leaveMessage) {
         // Paper end - Fix kick event leave message not being sent
+        com.kiocg.server.tasks.BossBarTask.removeFromAll(player.getBukkitEntity()); // KioCG - bossbar task
+        com.kiocg.server.commands.TickingViewerCommand.onPlayerQuit(player); // KioCG - ticking viewer
+
         ServerLevel level = player.level();
         player.awardStat(Stats.LEAVE_GAME);
         // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
@@ -639,7 +_,7 @@
         player.connection.internalTeleport(player.getX(), player.getY(), player.getZ(), player.getYRot(), player.getXRot()); // Paper
         player.connection.send(new ClientboundSetDefaultSpawnPositionPacket(level.getRespawnData()));
         player.connection.send(new ClientboundChangeDifficultyPacket(levelData.getDifficulty(), levelData.isDifficultyLocked()));
-        player.connection.send(new ClientboundSetExperiencePacket(player.experienceProgress, player.totalExperience, player.experienceLevel));
+        player.connection.send(new ClientboundSetExperiencePacket(player.insomniaProgress, player.totalExperience, player.experienceLevel)); // KioCG - EXP bar shows insomnia progress
         this.sendActivePlayerEffects(player);
         this.sendLevelInfo(player, level);
         this.sendPlayerPermissionLevel(player);
@@ -1023,7 +_,7 @@
         player.refreshEntityData(player); // CraftBukkit - SPIGOT-7218: sync metadata
         player.connection.send(new ClientboundSetHeldSlotPacket(player.getInventory().getSelectedSlot()));
         // CraftBukkit start - from GameRules
-        int i = player.level().getGameRules().get(GameRules.REDUCED_DEBUG_INFO) ? 22 : 23;
+        int i = player.level().getGameRules().get(GameRules.REDUCED_DEBUG_INFO) && !player.getBukkitEntity().hasPermission("kiocg.reduced_debug_info.bypass") ? 22 : 23; // KioCG - more permission
         player.connection.send(new ClientboundEntityEventPacket(player, (byte) i));
         float immediateRespawn = player.level().getGameRules().get(GameRules.IMMEDIATE_RESPAWN) ? 1.0F: 0.0F;
         player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.IMMEDIATE_RESPAWN, immediateRespawn));
